### PR TITLE
[ROCm] add Softmax Tunable Op

### PR DESCRIPTION
**Pull request by PeixuanZuo**
### Description
Add Softmax Tunable Op, only include blockwise vec implementation and composable kernel.
Related PR: https://github.com/microsoft/onnxruntime/pull/14475, https://github.com/microsoft/onnxruntime/pull/14612

### CODE REVIEW

**Comment by PeixuanZuo** [comment]
the performance of softmax with  stable diffusion softmax input
'''
python softmax_test.py 65536 4096 0 float16 --sort
SoftmaxTunable                                                                                                 float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   994.02 us, 1080.20 GB/s
DeviceReduceSoftmax<4,1,256,M_C1_S1,K_C256_S16,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>          float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   994.61 us, 1079.56 GB/s
DeviceReduceSoftmax<4,1,256,M_C1_S1,K_C256_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>           float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1466.23 us, 732.32 GB/s
DeviceReduceSoftmax<4,1,256,M_C2_S1,K_C128_S32,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>          float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1527.93 us, 702.75 GB/s
DeviceReduceSoftmax<4,1,256,M_C2_S1,K_C128_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>           float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1719.11 us, 624.59 GB/s
DeviceReduceSoftmax<4,1,256,M_C2_S1,K_C128_S16,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>          float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1829.03 us, 587.06 GB/s
SoftmaxBlockwise_8                                                                                             float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1845.70 us, 581.75 GB/s
SoftmaxBlockwiseStaticSelection                                                                                float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1846.13 us, 581.62 GB/s
SoftmaxBlockwise_16                                                                                            float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1869.16 us, 574.45 GB/s
DeviceReduceSoftmax<4,1,256,M_C4_S1,K_C64_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>            float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   1911.22 us, 561.81 GB/s
DeviceReduceSoftmax<4,1,256,M_C1_S1,K_C256_S32,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>          float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   2145.54 us, 500.45 GB/s
SoftmaxBlockwise_4                                                                                             float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   2204.02 us, 487.17 GB/s
DeviceReduceSoftmax<4,1,256,M_C8_S1,K_C32_S8,InSrcVectorDim_1_InSrcVectorSize_8_OutDstVectorSize_8>            float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   2412.08 us, 445.15 GB/s
SoftmaxBlockwise_2                                                                                             float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   2414.19 us, 444.76 GB/s
SoftmaxBlockwise_1                                                                                             float16 batch_count=65536 softmax_elements=4096 is_log_softmax=0   3803.88 us, 282.28 GB/s
...
'''

File: onnxruntime/core/providers/rocm/math/softmax_tunable_op.cuh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+#ifdef USE_COMPOSABLE_KERNEL
+    for (auto&& [_, op] : GetCKSoftmaxTypeStringAndOps<input_t, output_t, acc_t>()) {
+      ORT_UNUSED_PARAMETER(_);
+      this->RegisterOp(std::move(op));
+    }
+#endif  // USE_COMPOSABLE_KERNEL
+
+    // NOTE: the 1st kernel is SoftmaxBlockwise Original implementation.
+    this->SetDefaultId(0);
```
Comment (abudup) [inline onnxruntime/core/providers/rocm/math/softmax_tunable_op.cuh:74]: > `this->SetDefaultId(0);`

Is this necessary? If left unspecified, I believe that `0` is automagically the default implementation that is chose.

**Review by abudup — COMMENTED** [review]
Looks good, generally. Please address my comments.

File: onnxruntime/core/providers/rocm/math/softmax_ck.cuh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#include <utility>
+#include <vector>
+
+#ifdef USE_COMPOSABLE_KERNEL
+#include "ck/ck.hpp"
+#include "ck/library/tensor_operation_instance/gpu/softmax.hpp"
+#include "ck/tensor_operation/gpu/device/device_softmax.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+#endif
```
[the remark this answers is not in the document]
Comment (PeixuanZuo) [reply onnxruntime/core/providers/rocm/math/softmax_ck.cuh]: Done.

File: onnxruntime/core/providers/rocm/math/softmax_common.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "core/common/status.h"
+#include "core/providers/rocm/miopen_common.h"
+#include "core/providers/rocm/tunable/rocm_tunable.h"
+
+namespace onnxruntime {
+namespace rocm {
+
+template <typename input_t, typename output_t>
+struct SoftmaxParams : onnxruntime::rocm::tunable::OpParams {
```
[the remark this answers is not in the document]
Comment (PeixuanZuo) [reply onnxruntime/core/providers/rocm/math/softmax_common.h]: Done.

File: onnxruntime/core/providers/rocm/math/softmax_tunable_op.cuh (hunk 1 above)
Comment (PeixuanZuo) [reply onnxruntime/core/providers/rocm/math/softmax_tunable_op.cuh]: Yes, it's not necessary here.

**Review by abudup — APPROVED** [review]

**Pull request merged**